src/actor/calc.rs

```diff
@@ -0,0 +1,301 @@
+pub mod lookup;
+use super::stat::{SpecialStat, Stat, Stats};
+use crate::sim::SimState;
+use math::round::floor;
+
+#[allow(dead_code)]
+#[derive(Debug, Copy, Clone)]
+pub enum AttackType {
+    PHYSICAL,
+    MAGIC,
+}
+
+impl Default for AttackType {
+    fn default() -> Self {
+        AttackType::PHYSICAL
+    }
+}
+
+// https://www.akhmorning.com/allagan-studies/how-to-be-a-math-wizard/shadowbringers/damage-and-healing/#direct-damage-d
+pub fn direct_damage(
+    sim: &SimState,
+    potency: i64,
+    job: lookup::Job,
+    stats: &Stats,
+    attack_type: AttackType,
+    multipliers: Vec<f64>,
+) -> i64 {
+    let fatk = attack_power(job, stats.get(Stat::AttackPower));
+    let fdet = determination(stats.get(Stat::Determination));
+    // https://www.akhmorning.com/allagan-studies/how-to-be-a-math-wizard/shadowbringers/damage-and-healing/#direct-damage-d
+    // D1 = ⌊ Potency × f(ATK) × f(DET) ⌋ /100 ⌋ /1000 ⌋
+    let d1 = ((potency * fatk * fdet) / 100) / 1000;
+
+    let ftnc = tenacity(stats.get(Stat::Tenacity));
+    let wd = match attack_type {
+        AttackType::PHYSICAL => stats.get(Stat::PhysicalWeaponDamage),
+        AttackType::MAGIC => stats.get(Stat::MagicWeaponDamage),
+    };
+    let fwd = weapon_damage(job, wd);
+    // D2 = ⌊ D1 × f(TNC) ⌋ /1000 ⌋ × f(WD) ⌋ /100 ⌋ × Trait ⌋ /100 ⌋
+    let d2 = (((((d1 * ftnc) / 1000) * fwd) / 100) * job.trait_multiplier()) / 100;
+
+    let crit = critical_hit(
+        sim,
+        stats.get(Stat::CriticalHitRate),
+        stats.get_special(SpecialStat::CriticalHitPercentOverride),
+    );
+    let dh = direct_hit(sim, stats.get(Stat::DirectHitRate));
+    // D3 = ⌊ D2 × CRIT? ⌋ /1000 ⌋ × DH? ⌋ /100 ⌋
+    let d3 = (((d2 * crit) / 1000) * dh) / 100;
+    // D = ⌊ D3 × rand[95,105] ⌋ /100 ⌋
+    let d = d3 * sim.rng.random_from_range(95, 106) / 100;
+
+    // ⌊ ⌊ D × buff_1 ⌋ × buff_2 ⌋
+    multipliers
+        .iter()
+        .fold(d as f64, |total, multiplier| floor(total * *multiplier, 0)) as i64
+}
+
+// Level 80 F(AP)

```

Comment:
Comments like this should be doc comments whenever you can manage it (using `///`), so then they're more easily accessible to end users and in IDE highlights.